### ISSUE TITLE
determine_windows_extra_paths() returns list of dependency paths in random order, breaks building glib

### ISSUE THREAD

**Issue by lazka** [mesonbuild/meson#12330]
glib is currently not buildable in case glib is already installed on the system, this is due to some glib build commands trying to link to system glib during the build which is missing some new symbols. See downstream issue: https://gitlab.gnome.org/GNOME/glib/-/issues/3114

A bit of debugging shows that the PATH env var passed to commands (serialized via "extra_paths") is in a different order on every configure run. Looking at the function filling it shows that it uses unordered sets for everything and mixes build and system dependencies in multiple places and puts them into the same set(): https://github.com/mesonbuild/meson/blob/179355bf18f3c4216e8bb820fc19567e2a4bb640/mesonbuild/backend/backends.py#L1159

So sometimes the system DLL is in PATH before the meson built one.

Site note: whatever the fix is here, avoiding set() with deterministic output would help hunt down future issues in that code

**Comment by lazka** [comment]
@bruchar1 since you touched this last. Do you have any thoughts on how to best fix this?

**Comment by dcbaker** [comment]
@lazka I think you can just change `result: T.Set[str] = set()` to `result: mesonlib.OrderedSet[str] = mesonlib.OrderedSet()` and be good to go

**Comment by xclaesse** [comment]
@lazka OOC, what's the expected order? builddir or system first?

In related news, in https://github.com/mesonbuild/meson/pull/12059 I rework a lot that function, but I didn't pay attention to ordering, so that code needs revisiting too.

> avoiding set() with deterministic output would help hunt down future issues in that code

That's a fair point, I personally thought that set() was ordered the same way dict now are. That's a huge trap, we probably should do a project wide search-and-replace to never use it!

**Comment by eli-schwartz** [comment]
There are many, many, many cases where sets are entirely correct to use because order doesn't matter. Why would we ever want to replace those?

There are a bunch of cases where we know order matters and we use mesonlib.OrderedSet, again, as we should. There is apparently at least one case where order matters and we did NOT use mesonlib.OrderedSet, and we can fix that on its own.

**Comment by xclaesse** [comment]
> Why would we ever want to replace those?

OOC Is there a reason no to? Unstable order bit us many times for reproducibility, it's often far from obvious that order actually matters.

**Comment by eli-schwartz** [comment]
- suggesting that the entire python programming language is advised to give up the use of one of its most basic builtin datatypes certainly sounds novel to me :)
- import order circularity
- many times it is *very* obvious that order does not matter
- OrderedSet is slow. It's a replacement for one of the most basic language builtins, and that replacement is written in pure python instead of C. It cannot help but be comparatively slow :) and I would generally prefer to make meson faster, not slower. At the very least, I would want to vet every single proposed replacement to ensure either one of two things is true:
  - order matters in that specific case, in which case discussion over, we did in fact need it
  - it's not in a hot path and/or it's never used a large number of times, so even if we become slower, we don't become *noticeably* slower in the common case

**Comment by xclaesse** [comment]
It's tangential, I'm not really proposing to do it here. But wouldn't that be just a dict with None value? That would have pretty similar perf, no? To be honest I wrongly assumed that set() was actually a dict and was ordered the same way.

**Comment by xclaesse** [comment]
The reason I'm saying this is because it's not only about args order, it's also anything that gets pickled into coredata or any other president state that would cause spurious reconfigure. We felt into that trap many times already.

**Comment by dcbaker** [comment]
I did a few tests with `timeit`, `OrderedSet` is roughly 2-4x slower than `set` for common operations (initializtion, add, membership tests), and even slower than `list` in some cases (adding, but faster for membership tests). So I think it would be a bad idea to replace all `set` uses with `OrderedSet`

**Comment by xclaesse** [comment]
Wondering how you get such huge difference. There is some variability, set() and dict() always have similar perf, but wrapping dict into a class does have a 10-15% cost.

https://bpa.st/CXEQ
```
$ python set.py
set()		 1.0711724859429523
OrderedSet()	 1.2292896250728518
MySet()		 1.1717969930032268
dict()		 1.0361316660419106
```

Anyway, that's offtopic, sorry.

**Comment by dcbaker** [comment]
I should have been clearer, initialization with values. so something like `set(range(10))` vs `OrderedSet(range(10))`.

**Comment by xclaesse** [comment]
Right, that's indeed the worst case scenario. Note that our implementation can be improved, `dict({i: None for i in range(10)})` is faster. But still about 2x slower than set() initialization :/

**Comment by lazka** [comment]
> @lazka OOC, what's the expected order? builddir or system first?

Since there is no guarantee with system dependencies that they wont inject DLLs that conflict with other dependencies I'd say sort all system dependencies (anything not in the meson build dir) at the end.
